Test with vitest and pngjs

lubianat opened the pull request:
Building upon Will's 
- https://github.com/BioNGFF/ome-zarr.js/pull/19 
on a new PR

In relation to that, I've experimented on extracting out some parts of renderImage and on using https://www.npmjs.com/package/pngjs for testing without a browser

I've rolled back on JSdom for that, but we can go that path anyways


Will added "next step would be for you to 
- [x] open a PR directly against the main repo and 
- [x] add github action to run the tests for open PRs etc. "

Fixes #18.

will-moore committed:
Add simple test for renderThumbnail()

will-moore committed:
try to use jsdom canvas for testing without a browser

lubianat committed:
Extract out rbgData generation and DataUrl conversion

lubianat committed:
Use pngjs for testing in node

- Uses https://www.npmjs.com/package/pngjs

- Separates rendering of data url in DOM/browser (canvas)
from rendering in node/testing (pngjs)

- For testing, data URLs are different; using decoded pixel bytes

lubianat committed:
Merge branch 'test_with_vitest' into test_with_vitest_and_pngjs

lubianat committed:
Merge branch 'BioNGFF:main' into test_with_vitest_and_pngjs

lubianat committed:
Add CI workflow

lubianat committed:
Update package.json and tests to pass

- Setting longer timeout, as tests make calls that can take a few secs
- Updating package.json and tsconfig.json to account for the new parts

lubianat committed:
Merge branch 'test_with_vitest_and_pngjs' of https://github.com/lubianat/ome-zarr.js into test_with_vitest_and_pngjs

will-moore commented:
Tests are running 👍 
But failing:
```
npm error `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
```

lubianat committed:
Re-create the package-lock.json to fix inconsistencies

Note: running npm install with my npm version normalized package-lock.json and removed per-package license fields;

happy to adjust if the project prefers a specific npm version / lockfile format.

lubianat commented:
@lubianat tried fixing it here; workflow re-running needs approval

will-moore commented:
Build is green and tests passed 👍  Thanks...

brb..

will-moore commented:
Sorry - just getting back to this...

I realise this is going to conflict with subsequent PRs because of the formatting changes.
To mitigate that it would help if this was all formatted with e.g. `Prettier` which is what I typically use (I guess that should be automated at some point), or possibly reverted to just leave the essential code changes.

Also, there's a couple of functions without return types defined.

It would be nice to add a few more thumbnails to the test. The list of URLs on the index.html page (or at https://github.com/BioNGFF/ome-zarr.js/blob/main/docs/render-thumbnail.md?plain=1#L33) should provide a good mix of versions and other feature coverage, but the full test coverage can come in a follow-up PR.

Thanks

lubianat commented:
Only worry with more coverage is the test duration, as they depend on the network. 

Perhaps two test sets, one quick for local checks and a more complete for e.g. PRs 

But for now, I'll format with Prettier and add return types

lubianat committed:
Add Prettier formatting

lubianat commented:
@will-moore not sure on the correct Prettier settings to use; I tried to keep changes minimal. But whatever you prefer/use, we can set it up so we avoid running into that later?

lubianat committed:
Add some return types to functions

lubianat commented:
Also... I was reading a bit more and vitest has https://vitest.dev/guide/browser/, so it may make sense to add browser testing too. 

I think it makes sense to have a rendering setup that is somewhat independent from a browser, so maybe having both tests make sense?

lubianat committed:
Support vitest in browser

lubianat commented:
(browser testing still missing from GitHub Actions, will check how to do it)

will-moore commented:
Thanks for those changes...

I had a look at the vite browser testing (before you got node to use png lib instead for the data-Url).

However, I think it's probably a bit more hassle than it's worth, since ome-zarr.js isn't a UI application. If we know that it generates thumbnails without errors then we're pretty-much done.

will-moore commented on src/render.ts:
```diff
@@ -92,97 +97,171 @@ export async function renderThumbnail(
+): Promise<{
+  ndChunks: any[];
+  rbgData: Uint8ClampedArray;
+}> {
```
I was thinking that maybe this function could return just the `rgbData` (actually, this is rgba data) and `width, height` instead of the `ndChunks`. The chunks are only used for width and height. This would provide a simpler API for any external users and is easier to type (no `any[]`).
Maybe `function getRgba()` could return a single object:
```
{ data: Uint8ClampedArray, width: number, height: number }
```
This is kinda equivalent to https://developer.mozilla.org/en-US/docs/Web/API/ImageData (except ImageData is only in the browser, not in node.js).

lubianat committed:
Fix typescript ci errors

lubianat committed:
Change API for getRgba

will-moore reviewed (approved):
Looks good, thx
